lib/function.g

```diff
@@ -159,6 +159,31 @@ BIND_GLOBAL( "TYPE_OPERATION",
 ##
 DeclareOperationKernel("NameFunction", [IS_OBJECT], NAME_FUNC);
 
+#############################################################################
+##
+#O  NameFunction( <func> )  . . . . . . . . . . . . . . .  name of a function
+##
+##  <#GAPDoc Label="HasNameFunction">
+##  <ManSection>
+##  <Oper Name="HasNameFunction" Arg='func'/>
+##
+##  <Description>
+##  returns if a function has an assigned name.
+##  <P/>
+##  <Example><![CDATA[
+##  gap> NameFunction(SylowSubgroup);
+##  "SylowSubgroup"
+##  gap> HasNameFunction(SylowSubgroup);
+##  true
+##  gap> HasNameFunction(x->x);
+##  true
+##  ]]></Example>
+##  </Description>
+##  </ManSection>
+##  <#/GAPDoc>
+##
+##
+DeclareOperationKernel("HasNameFunction", [IS_OBJECT], HAS_NAME_FUNC);
```

Comment:
Huhm... So along with `SetNameFunction`, this "operation" now kind of is an attribute... Perhaps we should turn it into one (a custom one, with custom storage, of course) ? Then people who want to implement custom objects just need to implement that attribute in some way. It would perhaps also remove the need to document `HasNameFunction` (and `SetNameFunction`)?

Of course on a technical level, actually doing that might be hard. So I am not asking that this should be done in this PR; I am just thinking aloud. Perhaps @stevelinton has some insights on this, too?

Reply:
I'll be honest, kernel attributes/methods is one of those things I do not really understand. I'm happy if someone can point me to something similar I can cut+paste

Reply:
Also it's a bit weird, as we return "unknown" even if HasNameFunction is false.

Reply:
Also it's a bit weird, as we return "unknown" even if HasNameFunction is false.

Reply:
That's of course a valid point. But such behavior is not completely unheard of -- one can disable (both temporarily and permanently) "attribute storing" for select attributes or objects...

Of course I'd also be happy to change it to trigger an error if `NameFunction` is called when there is no name set, but then we'll likely have to adjust a bunch of code for that...

Reply:
Anyway, this attribute would also have to be a "mutable attribute", seeing as the name can be changed later on.

I've not yet had time to look deeper into this, but a patch could start out like this (completely untested, and clearly incomplete, as e.g. `TYPE_FUNCTION_WITH_NAME` still needs to be defined in the library):
```patch
diff --git a/lib/function.g b/lib/function.g
index a48381e29..de02482c9 100644
--- a/lib/function.g
+++ b/lib/function.g
@@ -157,7 +157,7 @@ BIND_GLOBAL( "TYPE_OPERATION",
 ##  <#/GAPDoc>
 ##
 ##
-DeclareOperationKernel("NameFunction", [IS_OBJECT], NAME_FUNC);
+DeclareAttributeKernel("NameFunction", IS_OBJECT, NAME_FUNC);
 
 
 #############################################################################
@@ -174,7 +174,7 @@ DeclareOperationKernel("NameFunction", [IS_OBJECT], NAME_FUNC);
 ##  </Description>
 ##  </ManSection>
 ##
-DeclareOperationKernel( "SetNameFunction", [IS_OBJECT, IS_STRING], SET_NAME_FUNC );
+InstallMethod( "SetNameFunction", [IsFunction, IS_STRING], SET_NAME_FUNC );
 
 
 #############################################################################
diff --git a/src/calls.c b/src/calls.c
index d2694d39b..f45959de2 100644
--- a/src/calls.c
+++ b/src/calls.c
@@ -965,14 +965,19 @@ Obj ArgStringToList(const Char *nams_c) {
 */
 static Obj TYPE_FUNCTION;
 static Obj TYPE_OPERATION;
+static Obj TYPE_FUNCTION_WITH_NAME;
+static Obj TYPE_OPERATION_WITH_NAME;
 
 static Obj TypeFunction(Obj func)
 {
-    return ( IS_OPERATION(func) ? TYPE_OPERATION : TYPE_FUNCTION );
+    // TODO: return correct type depending on whether NAME_FUNC is set or not
+    if (NAME_FUNC(func) == 0)
+        return (IS_OPERATION(func) ? TYPE_OPERATION : TYPE_FUNCTION);
+    else
+        return (IS_OPERATION(func) ? TYPE_OPERATION_WITH_NAME : TYPE_FUNCTION_WITH_NAME);
 }
 
 
-
 /****************************************************************************
 **
 *F  PrintFunction( <func> )   . . . . . . . . . . . . . . .  print a function
@@ -1224,12 +1229,12 @@ static Obj FuncCALL_FUNC_LIST_WRAP(Obj self, Obj func, Obj list)
 
 /****************************************************************************
 **
-*F  FuncNAME_FUNC( <self>, <func> ) . . . . . . . . . . .  name of a function
+*F  AttrNAME_FUNC( <self>, <func> ) . . . . . . . . . . .  name of a function
 */
-static Obj NAME_FUNC_Oper;
+static Obj NameFuncAttr;
 static Obj SET_NAME_FUNC_Oper;
 
-static Obj FuncNAME_FUNC(Obj self, Obj func)
+static Obj AttrNAME_FUNC(Obj self, Obj func)
 {
     Obj                 name;
 
@@ -1659,6 +1664,18 @@ static StructGVarFilt GVarFilts [] = {
 };
 
 
+/****************************************************************************
+**
+*V  GVarAttrs . . . . . . . . . . . . . . . . .  list of attributes to export
+*/
+static StructGVarAttr GVarAttrs [] = {
+
+    GVAR_ATTR(NAME_FUNC, "func", &NameFuncAttr),
+    { 0, 0, 0, 0, 0 }
+
+};
+
+
 /****************************************************************************
 **
 *V  GVarOpers . . . . . . . . . . . . . . . . .  list of operations to export
@@ -1722,6 +1739,8 @@ static Int InitKernel (
     /* install the type functions                                          */
     ImportGVarFromLibrary( "TYPE_FUNCTION",  &TYPE_FUNCTION  );
     ImportGVarFromLibrary( "TYPE_OPERATION", &TYPE_OPERATION );
+    ImportGVarFromLibrary( "TYPE_FUNCTION_WITH_NAME",  &TYPE_FUNCTION_WITH_NAME  );
+    ImportGVarFromLibrary( "TYPE_OPERATION_WITH_NAME", &TYPE_OPERATION_WITH_NAME );
     TypeObjFuncs[ T_FUNCTION ] = TypeFunction;
 
     /* init filters and functions                                          */
```

Reply:
While I agree this might have been a better idea from the start, I'd prefer not to change NameFunction to start returning errors, when previously it's always returned a valid string.

Reply:
Actually, that's not necessary at all: Even right now, calling `FuncNAME_FUNC` actually sets the name of the function explicitly -- i.e., this works just like regular attributes work: if no value is set, we compute one on the fly and store it.